core/schema/module.go

```diff
@@ -658,6 +672,205 @@ func (s *moduleSchema) moduleServe(ctx context.Context, modMeta *core.Module, ar
 	return void, query.ServeModule(ctx, modMeta, includeDependencies)
 }
 
+type moduleCallArgs struct {
+	Object   string
+	Function string
+	Parent   core.JSON
+	Inputs   core.JSON
+}
+
+func (s *moduleSchema) moduleCall(ctx context.Context, mod *core.Module, args moduleCallArgs) (inst dagql.Result[core.JSON], _ error) {
+	typeDef := &core.TypeDef{
+		Kind:     core.TypeDefKindObject,
+		AsObject: dagql.NonNull(&core.ObjectTypeDef{Name: args.Object}),
+	}
+	modType, ok, err := mod.ModTypeFor(ctx, typeDef, false)
+	if err != nil {
+		return inst, err
+	}
+	if !ok {
+		return inst, fmt.Errorf("failed to find receiver mod type %q", args.Object)
+	}
+
+	modObj := modType.(*core.ModuleObjectType)
+	callable, err := modObj.GetCallable(ctx, args.Function)
+	if err != nil {
+		return inst, err
+	}
+
+	var fields map[string]any
+	err = json.Unmarshal(args.Parent, &fields)
+	if err != nil {
+		return inst, fmt.Errorf("failed to unmarshal parent JSON: %w", err)
+	}
+	receiver, _, err := modType.ConvertFromSDKResult(ctx, nil, fields)
+	if err != nil {
+		return inst, fmt.Errorf("failed to decode parent object: %w", err)
+	}
+
+	dag, err := core.CurrentDagqlServer(ctx)
+	if err != nil {
+		return inst, fmt.Errorf("failed to get dag server: %w", err)
+	}
+
+	// HACK: need this so Host.directory IDs get loaded now and put in the cache,
+	// otherwise ones that haven't already been loaded in this engine won't
+	// be cached here and thus will result in the module function trying to
+	// load the dir from its own host in the container.
+	if err := loadIDsInServer(ctx, dag, modType, receiver); err != nil {
+		return inst, err
+	}
+
+	opts := &core.CallOpts{
+		ParentTyped:    nil,
+		ParentFields:   fields,
+		ParentModType:  modObj,
+		SkipSelfSchema: false,
+		Server:         dag,
+	}
+
+	dec := json.NewDecoder(bytes.NewReader(args.Inputs))
+	dec.UseNumber()
+	var callInputs map[string]any
+	err = dec.Decode(&callInputs)
+	if err != nil {
+		return inst, fmt.Errorf("failed to unmarshal inputs JSON: %w", err)
+	}
+	if dec.More() {
+		return inst, fmt.Errorf("unexpected extra data in inputs JSON")
+	}
+
+	for name, value := range callInputs {
+		argType, err := callable.ArgType(name)
+		if err != nil {
+			return inst, fmt.Errorf("failed to get type for argument %q: %w", name, err)
+		}
+		// NOTE: we use the SDK wire format here, even though this isn't really a SDK
+		// Mostly, this is just because it's easily available.
+		decodedValue, _, err := argType.ConvertFromSDKResult(ctx, nil, value)
+		if err != nil {
+			return inst, fmt.Errorf("failed to decode input %q: %w", name, err)
+		}
```

Comment:
This is where we use the new plumbing - it's not correct to use the `AnyResult` that we would have returned, because this is a manually created argument, there's no actual associated ID.

Side-note: why not actually use IDs? Well, that only works if it's all definitely in one engine, as soon as we start doing cross-engine calls, we need to send all the fields as part of the request - if we just send IDs, we'd have to replay the entire query, which is 1. wasted work, and 2. maybe incorrect if it's not deterministic.